docs/source/class_basics.rst

```diff
@@ -151,7 +151,101 @@ concrete. As with normal overrides, a dynamically typed method can
 implement a statically typed abstract method defined in an abstract
 base class.
 
+.. _protocol-types:
+
+Protocols and structural subtyping
+**********************************
+
+Mypy provides support for structural subtyping and protocol classes.
+To define a protocol class, one must inherit the special ``typing.Protocol``
+class:
+
+.. code-block:: python
+
+   from typing import Protocol
+
+   class SupportsClose(Protocol):
+       def close(self) -> None:
+          ...
+
+   class UnrelatedClass:
+       # some methods
+       def close(self) -> None:
+          self.resource.release()
+
+   def close_all(things: Sequence[SupportsClose]) -> None:
+       for thing in things:
+           thing.close()
+
+   close_all([UnrelatedClass(), open('some/file')])  # This passes type check
+
+Subprotocols are also supported. Inheriting from an existing protocol does
```

Comment:
Can you discuss subprotocols, generic protocols and recursive protocols in separate subsections (with titles), and with separate example code, to make the structure more clear, and to make it easier to find the relevant part in the documentation?